### ISSUE TITLE
Document which overheads are counted and which are excluded

### ISSUE THREAD

**Issue by teor2345** [privcount/privcount#191]
The stream event is also incosistent: it includes BEGINDIR (ORPort) directory requests, but not HTTP (DirPort) directory requests.

We could fix this by excluding BEGINDIR connections from all events and byte totals.
This seems to be the only plausible way to do this, because otherwise we risk the byte totals containing some directory request data and some exit request data.

**Comment by teor2345** [comment]
(This means excluding any stream, circuit, or connection that only had directory streams on it.)

**Comment by teor2345** [comment]
This might not be a significant issue for our measurements on the public tor network, because clients typically choose guards for their directory connections, not exits.

**Comment by teor2345** [comment]
The branch for this should be based on #190.

**Comment by robgjansen** [comment]
Should we separate BEGINDIR into it's own 'type' and let the PrivCount side filter those out? Or is it more appropriate to just ignore those completely? Traffic model counting is attempting to capture all traffic flowing through Tor, so it may be important to include them for that case?

**Comment by teor2345** [comment]
> Should we separate BEGINDIR into it's own 'type' and let the PrivCount side filter those out?
This is possible but will take more time to implement.
> Or is it more appropriate to just ignore those completely?
We already ignore DirPort requests, do you want to include or exclude those?
> Traffic model counting is attempting to capture all traffic flowing through Tor, so it may be important to include them for that case?
Then we should add separate counters for BEGINDIR and DirPort, so we can count DNS, Data Transfer (Exit/Entry/Middle/Rendezvous), and Directory (DirPort + BEGINDIR) traffic. (We might also want to count ORPort traffic, which is Data + BEGINDIR.)

Here's a quick design:
* an enum with a set of categories
* a classification function that decides which category each stream belongs in
* a summation function that takes a byte/cell count amount and the arguments to the classification function (and that also handles overflow with a tor_assert_nonfatal_once())
* additional fields on each event for the 4 different data categories
* modify PrivCount to parse those fields, and select the correct fields for each PrivCount counter

This could take up to a week to code and test.

Here's a more ambitious design that I would do if we were going to continually add more categories:
* move all the data categories to the end of the event line
* tag each data category with a string
* modify PrivCount to parse these strings and create a list of items (and do a >= on the items on the line)
* modify the injector for the new time positions

This might take a day or two more.

I think this is worthwhile if you ever plan on adding more data categories, like overhead (cells for setup, sendme, padding, etc.) vs data. If Mike Perry's traffic padding is ever merged, we'll want to do this.

**Comment by teor2345** [comment]
Oh, we also want one other category: DirPort self-checks, that is, Exit connections to an IP and port listed in the consensus as a DirPort. (This won't catch self-checks from new relays, they appear in the consensus *after* they check their DirPorts.)

**Comment by teor2345** [comment]
Hmm, we could catch around 40% of DirPort self-checks by always counting port 9030 as a DorPort self-check. This would only result in a small number of false positives.

**Comment by teor2345** [comment]
> I wonder if they would outweigh verifiable DirPort self-checks

The best way to find this out might be to run a data collection, and compare the number of verifiable self-checks with the number of requests to port 9030.

We could also use the size of the requests, or store the IPs for up to an hour, then check if they appear in the next consensus.

I'm not sure if the number of requests is worth the effort. It may be much simpler just to add a single line to the documentation:

"We expect relay DirPort self-checks to be a very small proportion of overall traffic. We can detect DirPort self-checks when an existing relay restarts, but not for new relays or relays that are missing  from the previous consensus."

I'd rather focus on getting BEGINDIR and DirPorts right, and maybe doing verifiable DirPort self-checks.

**Comment by teor2345** [comment]
This tor ticket has a list of the DirPort entry points:
https://trac.torproject.org/projects/tor/ticket/21206

**Comment by teor2345** [comment]
I think the relay traffic model looks something like this:

Responder (Listening Ports):
* ORPort
  * Data Transfer
  * Directory Documents (BEGINDIR: DirPort over ORPort)
* DirPort

Initiator (Outbound Connections):
* Exit
  * DNS
  * Data Transfer
* Relay
  * DirPort Self-Checks (a 3-hop path to another Exit)
  * ORPort Self-Checks (a 3-hop path to self)
    * ORPort Reachability
    * Bandwidth Self-Checks
* Client
  * Directory Documents (a direct connection to our Directory Guards' DirPorts or ORPorts)
  * Preemptive OR Connections (3-hop paths to other Relays)

**Comment by teor2345** [comment]
# Tor Network Overhead

The Tor network transfers TCP streams. Exits connect Client streams to
arbitrary remote Internet services, and Hidden Services connect Client streams
to a particular service, which may or may not be Internet-accessible.

The Tor network also has the following overheads:

## Directory

Tor's directory documents are used by clients to locate relays and build
circuits.

### Relay Descriptor Uploads

Tor relays upload their descriptors to the directory authorities via a direct
HTTP POST.

### Voting and Consensus Signatures

Tor directory authorities vote and sign consensuses every hour via a direct
HTTP POST.

### Directory Downloads

Relays and clients download directory documents from directory mirrors, either
over a one-hop BEGINDIR connection, or an unencrypted HTTP GET via the DirPort.

### Bandwidth Authorities

Bandwidth authorities build 2-hop paths through a relay and an exit, and check
the speed of the connection.

## Client

### Preemptive Circuits

Clients keep a small number of circuits open to speed up application requests.

## Relay

### ORPort Self-Checks

Relays create a 2-hop path to their own ORPort, and wait until they receive
a remote connection.

### ORPort Bandwidth Checks

Relays create a 2-hop path to their own ORPort, and transfer cells to measure
their own bandwidth.

### DirPort Self-Checks

Directory mirrors create a 3-hop path to their own DirPort, and download their
own descriptor.

## Onion Routing

### Authentication

Tor sends CERTS and AUTH* cells to authenticate to relays.

### Metadata

Tor sends VERSIONS and NETINFO cells.

### Circuit Management

Tor sends CREATE* and DESTROY cells to manage circuits.

### Data Flow Management

Tor sends SENDME cells to request more data.

### Padding

Tor automatically drops [V]PADDING cells: upcoming versions of tor may send
these cells by default. Tor also drops RELAY_DROP cells.

### Data Cell Overhead

Each cell contains a variable-length header depending on the cell type.
RELAY_DATA cells typically have 16 bytes of overhead (out of 514 bytes).

## Exit

### DNS Resolution

Exits perform DNS resolution on hostnames supplied by clients. CONNECT requests
connect the stream to the remote server. RESOLVE requests return the IP address
to the client.

### Exit Mapping

Tor runs a service which connects to every exit and discovers its exit IP
addresses.

**Comment by teor2345** [comment]
@robgjansen, I'm stuck on this issue, because I don't know what PrivCount is actually trying to count with each of the counters.

I've spent a few hours documenting the kinds of event data that each relay has available to it, and the overheads in the tor network.

Can you please read the [https://github.com/teor2345/privcount/blob/doc-role-overhead/doc/TorEvents.markdown#tor-relay-roles Tor Relay Roles section], and then let me know for each role: Guard, Middle, Exit, and Directory Mirror:

Are the bytes counters meant to count all bytes from this relay role?
If Directory Mirrors are counted, should DirPort bytes be counted exactly like stream bytes?

Are the stream counters meant to count all streams on this relay role?
(Note that streams are only available to Exits and Directory Mirrors)
If Directory Mirrors are counted, should each GET/POST request over a DirPort connection be counted like a BEGINDIR stream?

Are the circuit counters meant to count all circuits on this relay role?
If Directory Mirrors are counted, what should DirPort requests have as their circuit?

Are the connection counters meant to count all connections on this relay role?
If Directory Mirrors are counted, should DirPort connections be counted exactly like ORPort connections?
What about HTTP scanners / DirPortFrontPage?

Can you please read the [https://github.com/teor2345/privcount/blob/doc-role-overhead/doc/TorNetworkOverhead.markdown TorNetworkOverhead document].

This is what we currently do for each event type:

DNS:
* excludes invalid or closed circuits
Stream and Stream Bytes:
* excludes all streams initiated at this relay
  * directory uploads
  * client-like activities (directory downloads, preemptive circuits, etc)
* excludes non-Exit streams
  * but includes BEGINDIR and DNS RESOLVE
* excludes DirPort bytes
Circuit:
* includes all OR circuits
* excludes all circuits initiated at this relay
Connection:
* includes all OR connections
* excludes DirPort connections

Which of these overheads do you want to exclude from which events (or counters)?
For each overhead, what is your priority? How much time do you want to spend on it?

**Comment by robgjansen** [comment]
In general, previously, we were primarily trying to count Tor usage from client applications that connect to Tor, not internal Tor overheads. The distinction isn't always as clear as we might want; for example, if an application issues a GET request, that might cause the Tor process to refresh some internal state like downloading new directory info. Also, we are counting circuits, which one might count as "overhead", but we count these as they are probably one of the more important things Tor does in terms of overhead. But primarily we were concerned with application usage.

Ideally, I would have a magic traffic generator application that would replace a client. We would count the things that we need to count so that the traffic generator could perfectly reproduce the behavior of an "average" client, or the behavior of a distribution of clients.

For this, we primarily need to know about streams: the number of streams, the amount of data sent on streams, stream destinations, how long to wait in between creating new streams. The destination URLs could be used by the traffic generator and Tor would automatically create the needed number of circuits to support them. However, because there are lots of options for stream destinations, our privacy requirements would make it difficult to collect counts of specific destinations with good accuracy. So we instead opted to count circuits and the number of streams on each circuit, so that a traffic generator application could still combine streams to circuits to roughly match Tor usage.

This isn't perfect. Ideally we would have counts of every single type of stream, connection, DNS vs BEGINDIR vs DirPort, etc, but counters with smaller expected counts will generally have less accuracy. So we first focused on macro stats.

That doesn't mean we may not want to change. Separating all of the different types of events on the Tor side of things as in your proposal above makes sense, because then the PrivCount side can decide which ones to ignore, which ones to combine, etc. However, doing so is not a v0.2.0 issue in my opinion. I think it does make sense as part of v0.2.1, where we are documenting the counters, etc.

**Comment by robgjansen** [comment]
I'm looking at the overhead first, and thinking about the minimal changes that I think should be made to v0.2.0. The following is for specific counters, but may not apply to the "bytes" counters that are needed for traffic modeling. I'll address that in a separate comment.

  + DNS should be excluded, but we should add/correct and verify this in v0.2.1 (we will need it for alexa counters)
  + Exit mapping can be excluded
  + All directory overheads can be excluded (BEGINDIR and DirPort uploads/downloads, voting, BW auths)
  + All relay self checks and bandwidth checks can be excluded
  + All cells that are sent on circuits that are not used for the above would be included. I think that includes other clients' preemptive circuits, and circuits that get client streams attached.
  + Circuits and streams initiated at a relay should not be counted by that relay.

Based on how we do things now, I think the only "bug" is that we count BEGINDIR and DNS RESOLVE streams and we should be excluding them. I think that what this issue was created to fix.

Let me know if I am wrong about any of this. It's certainly a lot to keep straight!

**Comment by robgjansen** [comment]
In terms of relay roles, we were previously only trying to count stats from entries and exits. And we count independent counters from each position, to avoid the problem of double counting. Again, this doesn't include the new bytes counters.

These are counted on entries:
```
cat privcount/data_collector.py  | grep -e "Entry" -e "CircuitCells" -e "ClientIP" | cut -d'"' -f2 | sort | grep -v "#"
```

```
CircuitCellsIn
CircuitCellsOut
CircuitCellsRatio
CircuitsActiveEntry
CircuitsAllEntry
CircuitsInactiveEntry
ClientIPCircuitsActive
ClientIPCircuitsInactive
ClientIPsActive
ClientIPsInactive
ClientIPsUnique
```

The rest are counted on exits. We have not previously counted on middle node positions or directory mirror positions. If we do, I think that would be a bug.

**Comment by robgjansen** [comment]
See also #228, which might help clarify where things are counted.

**Comment by robgjansen** [comment]
I think it is an open question if we want to count any of the other overheads in the [Overhead Document](https://github.com/privcount/privcount/blob/master/doc/TorNetworkOverhead.markdown). For example, we might want to expand on the DNS events for the alexa counting in v0.2.1. Otherwise, perhaps we shouldn't implement other items until we have a need for them.

**Comment by robgjansen** [comment]
OK, now for the bytes events. Again, we would like to count the events that we would need for a traffic generator to be able to produce traffic flows in a test Tor network that is as close as possible to what Tor is transferring. I think that means we should count bytes being sent on streams, and from exit nodes only to avoid double counting. Our traffic model code takes these bytes events and computes a probabilistic model of roughly how often the client should send up a request packet or receive a data packet.

For this, we probably want to include the bytes from all streams that are getting created by the client's application, like HTTP requests as well as DNS lookups I think. The total number of streams and circuits that get created could be counted alongside the traffic model events if desired.

We will then rely on Tor to recreate the overheads, like consensus downloads, etc. And since those are a small fraction of the traffic being transferred, I think it's acceptable if they end up not being perfectly accurate in our test network.

Let me know if I am missing something.

**Comment by robgjansen** [comment]
Oh, and any changes that need to be made for the bytes events should be made in v0.2.0, because the traffic model is the first thing we will be counting.

**Comment by teor2345** [comment]
> For this, we primarily need to know about streams

This could be an issue for hidden services (and middle relays): only the hidden service (and client) see the streams on their circuit. So we might just have to count cells/bytes (and model traffic) in these cases.

> The following is for specific counters, but may not apply to the "bytes" counters that are needed for traffic modeling. ...
> DNS should be excluded, but we should add/correct and verify this in v0.2.1 (we will need it for alexa counters)

DNS is excluded from both stream and byte counters.

> Exit mapping can be excluded

Exit mapping is currently included, because it looks just like any other stream. Unless we know the details of the remote servers used by exit mapping, we can't exclude it: even if we did know the remote servers, such an implementation is fragile. And I doubt exitmap wants their servers listed publicly.

> All directory overheads can be excluded (BEGINDIR and DirPort uploads/downloads, voting, BW auths)

BEGINDIR circuits are currently included: #190 fixes the stream flag on the tor side, #230 extends the flag to connections, circuits, and bytes, and excludes flagged events from PrivCount counters.

DirPort server activity is excluded and will remain so: this includes receiving votes.
DirPort client activity is excluded and will remain so: this includes submitting votes.
(Not that we monitor any authorities yet.)
BEGINDIR client activity is excluded because it originates here.

Middle relays can not exclude circuits for anonymous upload of hidden service descriptors and bridge descriptors. But we see no streams on those circuits.

Bandwidth authorities have the same issues as exit mapping: it is not possible to exclude them reliably. But the servers used are public in some cases. Do you want to try to exclude them? (I don't think this is worthwhile, given the fragility.)

> All relay self checks and bandwidth checks can be excluded

DirPort self-checks have the same issues as exit mapping: it is not possible to exclude them reliably. Some servers used are in the consensus, others only join the consensus after they check their DirPort. Do you want to try to exclude them? (I don't think this is worthwhile, given the low volume.)

All self checks and bandwidth checks that originate here are excluded.
ORPort self-checks can not be identified in the middle or at the remote end.

ORPort bandwidth checks can be identified at the remote end, but not in the middle. Do you want to try to exclude them? (We can do this by marking RELAY_DROP cells, see #231.)

> All cells that are sent on circuits that are not used for the above would be included. I think that includes other clients' preemptive circuits, and circuits that get client streams attached.

It also includes preemptive circuits from other relays.

> Circuits and streams initiated at a relay should not be counted by that relay.

But they are counted at other relays, except for BEGINDIR.

We also exclude circuits (but not streams or bytes) where prevIsRelay and nextIsRelay are both false. This happens on a middle where the entry is not in the consensus, or on a middle when the exit is not in the consensus. This will be fixed in #199. (It can't happen on Exits, because most Exits have AllowSingleHopExit set to 0 by default.)

**Comment by teor2345** [comment]
> We have not previously counted on middle node positions or directory mirror positions. If we do, I think that would be a bug.

Since you check relays using the consensus, there is a bug where relays that aren't in the consensus this relay has will look like clients. As will bridges.

Since you don't have an is_dir flag on circuits or connections, there is a bug where circuits and connections are counted for directory mirrors.

(See my last comment for details on both of these.)

**Comment by teor2345** [comment]
> we would like to count the events that we would need for a traffic generator to be able to produce traffic flows in a test Tor network that is as close as possible to what Tor is transferring. I think that means we should count bytes being sent on streams, and from exit nodes only to avoid double counting.

Entry positions can't see stream details, as the cells are encrypted: double counting can never be an issue for streams.
...

> For this, we probably want to include the bytes from all streams that are getting created by the client's application, like HTTP requests as well as DNS lookups I think.

Depending on the client application, some DNS lookups are a separate stream (RESOLVE), others are performed as the first step in transferring data (CONNECT). See above for details.

...
> We will then rely on Tor to recreate the overheads, like consensus downloads, etc. And since those are a small fraction of the traffic being transferred, I think it's acceptable if they end up not being perfectly accurate in our test network.

Directory overhead is about 1 - 1.5 Gbps of 100 Gbps.
https://metrics.torproject.org/dirbytes.html
https://metrics.torproject.org/bandwidth.html

**Comment by teor2345** [comment]
I have opened tickets for the outstanding issues in this ticket.
The remaining task here is to document the overheads, and whether they are included in any counters.

**Comment by robgjansen** [comment]
> Bandwidth authorities have the same issues as exit mapping: it is not possible to exclude them reliably. But the servers used are public in some cases. Do you want to try to exclude them? (I don't think this is worthwhile, given the fragility.)

I agree with you that this probably is not worthwhile.

> DirPort self-checks have the same issues as exit mapping: it is not possible to exclude them reliably. Some servers used are in the consensus, others only join the consensus after they check their DirPort. Do you want to try to exclude them? (I don't think this is worthwhile, given the low volume.)

Again, I agree.

> ORPort bandwidth checks can be identified at the remote end, but not in the middle. Do you want to try to exclude them? (We can do this by marking RELAY_DROP cells, see #231.)

I'm ambivalent. Like much of the other overheads, I think this is a small fraction of the total network traffic.

**Comment by teor2345** [comment]
This has been done for a while.

**Comment by teor2345** [comment]
Actually merged in 1.0.0.
